Deploy to okfn server for better https support

k-nut opened the pull request:
This pull request aims to solve #4 by deploying to a codefor.de server that we have full control over so that we can make sure that http works.
The goal was not to lose any of the comfort that we currently have with github pages where we do not have to do anything to deploy.

Therefore I created a setup on a virtual machine and a [fabric](http://www.fabfile.org/) deployment script which can be run on travis to automatically deploy the newest version whenever the tests pass.

## This works in the following way 

### On the server
- Nginx serves static files from `/var/www/newest/`
- `/var/www/newest` is symlinked to `/home/deploy/verisons/newest`
- The `/home/deploy/versions` folder contains the `newest` folder and one folder per deployed revision. The newest revision itself is symlinked to `newest` (and through that to `/var/www/newest` which nginx serves).

### With fabric
- We bundle all files in the repository into a `.tar.gz`
- This tarfile is uploaded to the server
- Is is extracted and placed under `/home/deploy/versions/<current_revision>`
- `/home/deploy/versions/<current_revision>` is symlinked to `/home/deploy/versions/newest`
and therefore set to be served by nginx

Doing all the symlink helps us to easily roll back to a former revision if anything ever goes wrong with a deployment. We just need to change where `/home/deploy/versions/newest` points and we are done.

## Problems & last steps needed
One problem here is that by creating an archive for the whole project we also include files like `.git` and all other files that do not have to be deployed. I therefore propose moving the `index.html`, assets, javascript etc. into an `app` folder and just deploying that.

As a last step we need to create a new ssh key for travis and add that to the server.

Todo
- [ ] move files to `app` directory (optional)
- [x] add deploy key to travis
- [x] add travis deploy key to `/home/deploy/.ssh/allowed_hosts`
- [x] change travis script to also deploy
- [x] change CNAME to point to new server

## 🚀 

I hope that I made clear how I see this working. 

I am happy to answer all questions and looking forward to your feedback!

torfsen commented on fabfile.py:
```diff
@@ -0,0 +1,38 @@
+def create_artifact(current_revision):
+    """ compress all files into a .tar.gz archive for uploading """
+    archive_path = '/tmp/{revision}.tar.gz'.format(revision=current_revision)
+    local('tar -czf {archive_path} --exclude=.git *'.format(archive_path=archive_path))
```
Do we need that `--exclude` if we use a separate `app` folder?

k-nut replied:
nope. In that case we could just create a `.tar.gz` for `app` and deploy that

johnjohndoe replied:
I voted for staying with the current folder structure. I think the deployment can just filter out the _.git/_ folder.

torfsen commented on fabfile.py:
```diff
@@ -0,0 +1,38 @@
+from fabric.api import run, env, local, put, cd
```
I'd love to see a (short) high-level comment on what's being done in the file.

torfsen commented on fabfile.py:
```diff
@@ -0,0 +1,38 @@
+
+def upload_artifact(revision):
+    """ upload the archive to the server and extract it """
+    remote_versions_path = '/home/deploy/versions'
```
Please move the shared path-prefixes (e.g. `/home/deploy`) into global constants to ease maintenance.

k-nut replied:
Done in 6a2275a

johnjohndoe commented:
@k-nut Please add some high level documentation to your pull request which describes the deployment. Please state what triggers the deployment, which parts are automated and when it happens.
Please add a "Deployment" section to bottom of the _README.md_. Dependent on the amount of documentation it might make sense to create a separate _DEPLOYMENT.md_ and link to it.

torfsen commented:
Thanks for the work, @k-nut! I'd prefer a simpler setup, but it seems that you've put some thoughts into making this as simple as possible.

One thing that's missing from my point of view is documentation for this setup and the deployment process as part of the repository itself. You can add it to the README or to a separate document, and it doesn't have be much -- basically what you've wrote in your PR comment in its final form.

By the way, what's the easiest way to test this PR?

k-nut commented:
@johnjohndoe @torfsen I updated the readme in 8d1bd35. Is this sufficient for you?

k-nut commented:
In order for travis to be able to deploy I changed the language to `python` which is needed for fabric. Travis still allows you to use `npm` in that case which is all we need. You can see an example of a build and deploy here: https://travis-ci.org/wo-ist-markt/wo-ist-markt.github.io/builds/149556355.

It looks to me though like one of our functions is failing because we are using the new ecmascript 6 `=>` syntax. I think we have to change this to old javascript if we want to make sure that the test works.

k-nut commented:
I hope that I addressed all of your points. Let me know if anything is still unclear / needs to be updated. 

@torfsen I think the best way to "test" this is to either see what is happening or travis or to run the deployment locally. For that you would need to give me your public ssh key so that I can add you to the server. That might be a good idea anyways so that I am not the only one who could fix things there.

johnjohndoe commented on README.md:
```diff
@@ -146,3 +146,16 @@ Tests can be run using [npm][npm]:
 [wuppertal-markets]:https://www.wuppertal.de/tourismus-freizeit/einkaufen/102370100000204430.php
+
+
+## Deployment
```
Please consolidate the wording - I recommend using `Travis CI` not `travis` or `Travis`.

k-nut replied:
✅ a61bff0fee3bc0ee4949272ac536fb04138c2060

johnjohndoe commented on README.md:
```diff
@@ -146,3 +146,16 @@ Tests can be run using [npm][npm]:
+under https://wo-ist-markt.de.
+
+The command that travis executes is `fab deploy -i markt_deploy_id_rsa -H deploy@kiesinger.okfn.de:2207`
+where markt_deploy_id_rsa is an ssh key to log into the server. It is commited to this repository in
```
Please format `where markt_deploy_id_rsa` with backticks.

k-nut replied:
So would you drop that paragraph entirely?

johnjohndoe replied:
After the latest problems I would leave the paragraph in there. Please format the `where markt_deploy_id_rsa` with backticks. Please replace the last sentence with: The command to deploy manually is: `fab deploy -H deploy@kiesinger.okfn.de:2207`.

johnjohndoe commented on .travis.yml:
```diff
@@ -1,4 +1,18 @@
+  - npm test
+
+after_success:
+  - test $TRAVIS_PULL_REQUEST == "false" && test $TRAVIS_BRANCH == "master" && fab deploy -i markt_deploy_id_rsa -H deploy@kiesinger.okfn.de:2207
```
A short but descriptive comment would be nice here.

k-nut replied:
Added comments in bfdaf605e2195c28b378d792fd696a5e796b3f38

johnjohndoe commented on .travis.yml:
```diff
@@ -1,4 +1,18 @@
+    - '2.7'
+
+before_install:
+    - openssl aes-256-cbc -K $encrypted_02f86fe65d47_key -iv $encrypted_02f86fe65d47_iv
```
A short but descriptive comment would be nice here.

k-nut replied:
Added comments in bfdaf605e2195c28b378d792fd696a5e796b3f38

johnjohndoe commented on fabfile.py:
```diff
@@ -0,0 +1,50 @@
+                                                          revision=revision))
+
+def list_versions():
+    run('ls {base}'.format(base=BASE_PATH))
```
not `ls -l`?

k-nut replied:
✅  	85acbf1

k-nut commented:
Just pushed another change that replaces the mentioned `=>` fatArrow syntax from the test so that travis can execute them properly. 876c2b2

torfsen commented:
Some minor issues, but otherwise looks great now!

k-nut commented:
So @johnjohndoe proposed rewriting this into fewer logical commits. Should I do it or do you think that this is sufficient? And if yes do you have a proposed way for resturcturing? Again, this might be a chance to use a [squash commit](https://github.com/blog/2141-squash-your-commits) maybe?

Or are there any other changes that you would like to see before merging?

k-nut commented:
I just rebased yesterday. So can we just use the squash commit feature to merge this?

johnjohndoe commented:
Ah, good. Either, we use the GitHub squash feature or you can do it yourself to edit the commit message as you want it to be. Are there additional changes required with regards to #125?

k-nut commented:
🔨 squashed, updated and rebased. I think this is good to 🚢

johnjohndoe commented:
:rocket: merged! Thanks a lot for this nice piece of work!
